x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java

```diff
@@ -250,42 +253,94 @@ public void onFailure(Exception e) {
 
     public static class DataFrameAnalyticsTask extends AllocatedPersistentTask implements StartDataFrameAnalyticsAction.TaskMatcher {
 
+        private final Client client;
+        private final ClusterService clusterService;
+        private final DataFrameAnalyticsManager analyticsManager;
         private final StartDataFrameAnalyticsAction.TaskParams taskParams;
         @Nullable
         private volatile Long reindexingTaskId;
+        private volatile boolean isStopping;
 
         public DataFrameAnalyticsTask(long id, String type, String action, TaskId parentTask, Map<String, String> headers,
+                                      Client client, ClusterService clusterService, DataFrameAnalyticsManager analyticsManager,
                                       StartDataFrameAnalyticsAction.TaskParams taskParams) {
             super(id, type, action, MlTasks.DATA_FRAME_ANALYTICS_TASK_ID_PREFIX + taskParams.getId(), parentTask, headers);
+            this.client = Objects.requireNonNull(client);
+            this.clusterService = Objects.requireNonNull(clusterService);
+            this.analyticsManager = Objects.requireNonNull(analyticsManager);
             this.taskParams = Objects.requireNonNull(taskParams);
         }
 
         public StartDataFrameAnalyticsAction.TaskParams getParams() {
             return taskParams;
         }
 
-        public void setReindexingTaskId(long reindexingTaskId) {
+        public void setReindexingTaskId(Long reindexingTaskId) {
             this.reindexingTaskId = reindexingTaskId;
         }
 
         @Nullable
         public Long getReindexingTaskId() {
             return reindexingTaskId;
         }
+
+        public boolean isStopping() {
+            return isStopping;
+        }
+
+        @Override
+        protected void onCancelled() {
+            stop(getReasonCancelled(), TimeValue.ZERO);
+        }
+
+        public void stop(String reason, TimeValue timeout) {
+            isStopping = true;
+            if (reindexingTaskId != null) {
+                cancelReindexingTask(reason, timeout);
```

Comment:
There seems to be a race condition here between the reindex finishing, and the task id being set to null. 

If the reindexing is finished, but we have yet to complete the `refresh` request, we still have the reindexId set, but there is no task matching that ID. This means that we would get a node failure error saying that the task could not be found. This `cancelReindexingTask` would then throw, and we would not call `analyticsManager.stop(this);`.

We may have to check the `cancelReindexResponse.getNodeFailures()` to verify it is not caused by a `resource_not_found_exception` or remove this race condition somehow

Reply:
I don't it's possible to avoid that race. I'll make sure we catch if the task is no longer there.